File: Linear Control Course/Chapter22/Lesson2/Chapter22_Lesson2.py
import numpy as np
import matplotlib.pyplot as plt
import control as ct  # python-control library

# Plant and controller parameters
Kp = 2.0           # plant DC gain
tau_p = 0.1        # plant time constant
Kc = 5.0           # proportional controller gain

# Define plant P(s) = Kp / (tau_p s + 1)
P = ct.TransferFunction([Kp], [tau_p, 1.0])

# Define controller C(s) = Kc
C = ct.TransferFunction([Kc], [1.0])

# Loop transfer L(s) and complementary sensitivity T(s)
L = C * P
T = ct.feedback(L, 1)  # feedback(L,1) = L / (1 + L)

# Bode magnitude of T(jw) to study noise shaping
w = np.logspace(0, 4, 400)  # 1 rad/s to 10^4 rad/s
mag_T, phase_T, omega = ct.bode_plot(T, w, Plot=False)

plt.figure()
plt.semilogx(omega, 20 * np.log10(mag_T))
plt.xlabel("Frequency (rad/s)")
plt.ylabel("|T(jw)| (dB)")
plt.title("Complementary Sensitivity Magnitude")
plt.grid(True)

# Approximate output noise variance for band-limited white noise
sigma_n2 = 1.0  # noise PSD level
Phi_n = sigma_n2 * np.ones_like(omega)
# Using a simple Riemann sum approximation:
sigma_y2 = (1.0 / (2.0 * np.pi)) * np.trapz((mag_T ** 2) * Phi_n, omega)
print("Approximate output noise variance due to sensor noise:", sigma_y2)

plt.show()

File: Tutorials/Codes/Linear_Control_Course/Chapter22/Lesson2/Chapter22_Lesson2.m
% Parameters
Kp    = 2.0;
tau_p = 0.1;
Kc    = 5.0;

% Plant and controller
s = tf('s');
P = Kp / (tau_p * s + 1);
C = Kc;

% Loop transfer and complementary sensitivity
L = C * P;
T = feedback(L, 1);  % T(s) = L / (1 + L)

figure;
bode(T);
grid on;
title('Complementary Sensitivity T(s)');

% Time-domain simulation in Simulink (outline):
% 1. Create a Simulink model with:
%    - Sum block for error e = r - y_meas
%    - Gain block C(s) (for proportional control just a constant)
%    - Transfer Fcn block for P(s)
%    - Sum block to form y_meas = y + n
%    - Noise source: Band-limited White Noise block feeding the measurement
% 2. Connect the blocks according to the conceptual diagram.
% 3. Use a Step block for reference r(t) and Scope blocks to observe y(t) and u(t).
% 4. Use the 'sim' command from MATLAB to run parameter sweeps on Kc and observe
%    changes in noise level vs tracking speed.
